File: server/src/collectors/msilUmishiru.js
/**
 * MSIL Umishiru (海しる) - JCG Maritime Domain Awareness public APIs
 * https://portal.msil.go.jp/apis
 * Requires subscription key (trial key available on portal); honour UMISHIRU_API_KEY env var.
 * Without a key we fall back to seed port positions.
 */

const TIMEOUT_MS = 10000;
const API_KEY = process.env.UMISHIRU_API_KEY;

const SEED_PORTS = [
  { name: '東京港', lat: 35.61, lon: 139.77 },
  { name: '横浜港', lat: 35.45, lon: 139.66 },
  { name: '名古屋港', lat: 35.08, lon: 136.88 },
  { name: '大阪港', lat: 34.65, lon: 135.43 },
  { name: '神戸港', lat: 34.68, lon: 135.18 },
  { name: '博多港', lat: 33.60, lon: 130.40 },
  { name: '那覇港', lat: 26.21, lon: 127.67 },
  { name: '苫小牧港', lat: 42.62, lon: 141.62 },
];

export default async function collectMsilUmishiru() {
  let features = [];
  let source = 'seed';
  if (API_KEY) {
    try {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
      const url = 'https://portal.msil.go.jp/api/v1/ports?bbox=122,24,146,46&limit=1000';
      const res = await fetch(url, { signal: controller.signal, headers: { 'Ocp-Apim-Subscription-Key': API_KEY } });
      clearTimeout(timer);
      if (res.ok) {
        const data = await res.json();
        const list = Array.isArray(data) ? data : (data?.features ?? data?.items ?? []);
        for (const p of list) {
          const lat = p?.lat ?? p?.geometry?.coordinates?.[1];
          const lon = p?.lon ?? p?.geometry?.coordinates?.[0];
          if (lat == null || lon == null) continue;
          features.push({
            type: 'Feature',
            geometry: { type: 'Point', coordinates: [Number(lon), Number(lat)] },
            properties: { name: p.name ?? p.properties?.name ?? null, source: 'msil_umishiru' },
          });
        }
        if (features.length) source = 'live';
      }
    } catch { /* fall through to seed */ }
  }
  if (features.length === 0) {
    features = SEED_PORTS.map(p => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [p.lon, p.lat] },
      properties: { name: p.name, source: 'msil_seed' },
    }));
  }
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source,
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      description: 'MSIL Umishiru ports (JCG MDA)',
      key_required: true,
    },
  };
}

File: native/collectors/transport/sources/msil_umishiru.c
/* collectors/transport/sources/msil_umishiru.c
 * Port of server/src/collectors/msilUmishiru.js. JCG MDA "Umishiru" ports API
 * (key-gated). With UMISHIRU_API_KEY → single bbox GET → Features. Without a
 * key the JS emits only the curated SEED, which rule 8 drops → 0 rows. */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include <stdio.h>
#include <stdlib.h>

#define API_URL "https://portal.msil.go.jp/api/v1/ports?bbox=122,24,146,46&limit=1000"

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *k = getenv("UMISHIRU_API_KEY");
  if (!k || !*k) {
    fprintf(stderr, "[msil-umishiru] gated (no UMISHIRU_API_KEY)\n");
    return 0;                                   /* JS: SEED-only → dropped */
  }

  char keyh[256];
  snprintf(keyh, sizeof keyh, "Ocp-Apim-Subscription-Key: %s", k);
  const char *hdrs[] = { keyh, NULL };
  cJSON *data = feed_get_json_h(ctx->http, API_URL, hdrs, 10000);
  if (!data) return -1;

  /* list = Array.isArray(data) ? data : (data?.features ?? data?.items ?? []) */
  cJSON *list = NULL;
  if (cJSON_IsArray(data)) list = data;
  else {
    cJSON *fe = cJSON_GetObjectItem(data, "features");
    if (cJSON_IsArray(fe)) list = fe;
    else {
      cJSON *it = cJSON_GetObjectItem(data, "items");
      if (cJSON_IsArray(it)) list = it;
    }
  }

  cJSON *features = cJSON_CreateArray();
  if (list) {
    cJSON *p;
    cJSON_ArrayForEach(p, list) {
      /* lat = p?.lat ?? p?.geometry?.coordinates?.[1] */
      cJSON *lat = cJSON_GetObjectItem(p, "lat");
      cJSON *lon = cJSON_GetObjectItem(p, "lon");
      cJSON *geom = cJSON_GetObjectItem(p, "geometry");
      cJSON *gc = geom ? cJSON_GetObjectItem(geom, "coordinates") : NULL;
      if ((!lat || cJSON_IsNull(lat)) && gc && cJSON_IsArray(gc))
        lat = cJSON_GetArrayItem(gc, 1);
      if ((!lon || cJSON_IsNull(lon)) && gc && cJSON_IsArray(gc))
        lon = cJSON_GetArrayItem(gc, 0);
      if (!lat || cJSON_IsNull(lat) || !lon || cJSON_IsNull(lon)) continue;

      cJSON *f = cJSON_CreateObject();
      cJSON_AddStringToObject(f, "type", "Feature");
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(cJSON_GetNumberValue(lon)));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(cJSON_GetNumberValue(lat)));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);

      cJSON *pr = cJSON_CreateObject();          /* EXACT JS key order */
      /* name = p.name ?? p.properties?.name ?? null */
      cJSON *nm = cJSON_GetObjectItem(p, "name");
      if (!nm || cJSON_IsNull(nm)) {
        cJSON *pp = cJSON_GetObjectItem(p, "properties");
        nm = pp ? cJSON_GetObjectItem(pp, "name") : NULL;
      }
      cJSON_AddItemToObject(pr, "name",
        (nm && !cJSON_IsNull(nm)) ? cJSON_Duplicate(nm, 1) : cJSON_CreateNull());
      cJSON_AddStringToObject(pr, "source", "msil_umishiru");
      cJSON_AddItemToObject(f, "properties", pr);
      cJSON_AddItemToArray(features, f);
    }
  }
  cJSON_Delete(data);

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[msil-umishiru] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def msil_umishiru_def = {
  .id = "msil-umishiru", .collector = "transport",
  .name = "MSIL Umishiru Public APIs", .name_ja = "海しる 公開API",
   .update_interval_sec = 900, .run = run,
};
REGISTER_SOURCE(msil_umishiru_def)
